File: Difficulty/Hard/[188] Best Time to Buy and Sell Stock IV/solution.cpp
/*
 * @lc app=leetcode id=188 lang=cpp
 *
 * [188] Best Time to Buy and Sell Stock IV
 */

// @lc code=start
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        if (k <= 0 || prices.empty()) return 0;
        if (k >= prices.size() / 2)
            return greedy(prices);
        vector<vector<int>> dp(k + 1, vector<int>(2, 0));
        for (int i = 1; i <= k; i++) {
            dp[i][0] = 0;
            dp[i][1] = -prices[0];
        }
        for(int i = 1; i < prices.size(); i++){
            for(int j = 1; j <= k; j++){
                dp[j][0] = max(dp[j][0], dp[j][1] + prices[i]);
                dp[j][1] = max(dp[j][1], dp[j-1][0] - prices[i]);
            }
        }
        return dp[k][0];
    }
private:
    int greedy(const vector<int>& prices) {
        int res = 0;
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] > prices[i-1])
                res += prices[i] - prices[i-1];
        }
        return res;
    }
};
// @lc code=end

File: Difficulty/Hard/[188] Best Time to Buy and Sell Stock IV/solution.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        if k <= 0 or not prices: return 0
        if k >= len(prices) // 2:
            return self.greedy(prices)
        dp = [[0] * 2 for _ in range(k+1)]
        for i in range(1, k+1):
            dp[i][0] = 0
            dp[i][1] = -prices[0]
        for i in range(1, len(prices)):
            for j in range(1, k+1):
                dp[j][0] = max(dp[j][0], dp[j][1] + prices[i])
                dp[j][1] = max(dp[j][1], dp[j-1][0] - prices[i])
        return dp[k][0]
            
    def greedy(self, prices):
        res = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                res += prices[i] - prices[i - 1]
        return res